Kernel/Net/NetworkingManagement.cpp

```diff
@@ -93,19 +93,19 @@ ErrorOr<NonnullOwnPtr<KString>> NetworkingManagement::generate_interface_name_fr
     return name;
 }
 
-UNMAP_AFTER_INIT LockRefPtr<NetworkAdapter> NetworkingManagement::determine_network_device(PCI::DeviceIdentifier const& device_identifier) const
+UNMAP_AFTER_INIT ErrorOr<NonnullLockRefPtr<NetworkAdapter>> NetworkingManagement::determine_network_device(PCI::DeviceIdentifier const& device_identifier) const
 {
-    if (auto candidate = E1000NetworkAdapter::try_to_initialize(device_identifier); !candidate.is_null())
```

Comment:
We are still using the same `LockRefPtr` but skipping the `!adapter.is_null()` check. My C++ isn't good and most likely missing something obvious, but either it was unnecessary before or it should be added back? Sorry if I'm way off here..

Reply:
Wondered the same thing during the video :^). Maybe we could split this method into two: checking if the adapter matches the `PCI::DeviceIdentifier` and actually try to initializing it, the method currently tries to do both things. Then we could actually hard error on the initialization part. Rough sketch:

```
static ErrorOr<NonnullLockRefPtr<E1000ENetworkAdapter>> try_to_initialize(PCI::DeviceIdentifier const&);
static bool matches_device_identifier(PCI::DeviceIdentifier const&);
```
And in `NetworkingManagement::determine_network_device` the blocks could then look like this:

```
if(E1000ENetworkAdapter::matches_device_identifier(device_identifier))
        return TRY(E1000ENetworkAdapter.try_to_initialize(device_identifier));
```

Reply:
`LockRefPtr::operator bool()` is defined as `{ return !is_null(); }`, so `if (ptr)` will null-check.

Reply:
```
    NonnullRefPtr<T> release_nonnull()
    {
        auto* ptr = leak_ref();
        VERIFY(ptr);
        return NonnullRefPtr<T>(NonnullRefPtr<T>::Adopt, *ptr);
    }
```

It will never return `nullptr`. If it is a `nullptr` it will fall through to the next line.

Reply:
You are half right. It won't return `nullptr`... because the `VERIFY` means it will panic the kernel instead.

Reply:
Ahh, right, I get it now, this code is actually correct.I was reading it as "if the call succeeds with no error, then release nonnull", which would fail if the result is nullptr, which is a valid state. But it's not, it's actually two steps because of the TRY. This is the same logic, just spread out:

```cpp
auto candidate = TRY(E1000NetworkAdapter::try_to_initialize(device_identifier))
if (candidate)
  return candidate.release_nonnull();
```

The `if (candidate)` check will use `LockRefPtr::operator bool()`, and thus this code is actually behaving correctly.